README.md
#Django WhosHere 

[![Join the chat at https://gitter.im/Koed00/django-whoshere](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/Koed00/django-whoshere?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![](https://travis-ci.org/Koed00/django-whoshere.svg?branch=master)](https://travis-ci.org/Koed00/django-whoshere)[![Coverage Status] (https://coveralls.io/repos/Koed00/django-whoshere/badge.svg?branch=master)](https://coveralls.io/r/Koed00/django-whoshere?branch=master)

A lightweight Django Admin plugin showing who's logged in from where using the cache. 
Supports [Telize](https://www.telize.com), [django-ipware](https://github.com/un33k/django-ipware), [user-agents](https://github.com/selwin/python-user-agents)  and [GeoIP](https://docs.djangoproject.com/en/1.8/ref/contrib/gis/geoip/)

##Requirements

- [Django](https://www.djangoproject.com) >= 1.7
- [requests](https://github.com/kennethreitz/requests)

####Optional

- [user-agents](https://github.com/selwin/python-user-agents)  Adds nicer user agent formatting
- [django-ipware](https://github.com/un33k/django-ipware) More robust way of determining a users IP
- [GeoIP](https://docs.djangoproject.com/en/1.8/ref/contrib/gis/geoip/)  For geolocation instead of [Telize](https://www.telize.com).

##Installation
- Make sure you have Django's [Cache](https://docs.djangoproject.com/en/1.8/topics/cache/)  backend set up.
- Install using pip: `pip install django-whoshere`
- Add `django_whoshere` to `INSTALLED_APPS` in settings.py:
```python
INSTALLED_APPS = (
    # other apps
    'django_whoshere',
)
```
- Add `django_whoshere.middelware.TrackMiddleware` to your `MIDDLEWARE_CLASSES`. 
Make sure it comes after your Authentication middleware.

```python
MIDDLEWARE_CLASSES = (
    # other middleware
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
    'django_whoshere.middleware.TrackMiddleware',
   # more middleware
)
```  
####Optional
- Install  [user-agents](https://github.com/selwin/python-user-agents) for nicer user agent formatting
- Install [django-ipware](https://github.com/un33k/django-ipware) for a better way to determine real ip addresses.
- Configure [GeoIP](https://docs.djangoproject.com/en/1.8/ref/contrib/gis/geoip/)  or install [django-geoip-utils](https://github.com/Gidsy/django-geoip-utils)  for geo location instead of   [Telize](https://www.telize.com).

##Configuration
No configuration is needed but these settings are provided for convenience:

- `WHOSHERE_TIMEOUT = 300`
Sets the timeout for user activity. Defaults to 300 seconds.
- `WHOSHERE_LABEL = 'Active Users'`
Overrides the admin link label. Defaults to 'Active Users'
- `WHOSHERE_PREFIX = 'whoshere'`
Prefix used in cache keys. Defaults to 'whoshere'.
- `WHOSHERE_TELIZE = True`
You can set this to 'False' to turn off geolocation lookups with   [Telize](https://www.telize.com).
Is always off if GeoIp is installed.

##Template tags
Besides the admin interface, WhosHere's functions are also available as template tags:
```html+django
{% load whoshere %}

There are now {% active_user_count %} users logged in.

{% active_users as users %}
<ul>
{% for user in users %}
   <li>{{ user.username }} ({{ user.email }})</li>
{% endfor %}
</ul>

<p>Your IP address is {% your_ip %}</p>
<p>Your browser and platform is {% your_agent %}</p>
<p>You live in {% your_city %}, {% your_country %}</p>
```

##Geolocation
Originally this plugin started with GeoIP support only, but this means you have to install the rather large database that it comes with. 
Not a problem if you already use it for other things, but not a lightweight solution if you only use it for WhosHere.
As an alternative WhosHere uses the free  [Telize](https://www.telize.com) API over HTTPS by default to find the location of your logged in users.
This comes with some caveats:

- The [Telize](https://www.telize.com) API is [opensource](https://github.com/fcambus/telize), but you are sending users IP addresses, albeit anonymously, to a third party.
- Being a free API it can sometimes be unavailable or slow
- Your server needs HTTPS access to a remote location

You can turn off Telize lookups with `WHOSHERE_TELIZE =  False` in your `settings.py` or by installing and configuring GeoIP.

##Notes
- Middleware is kept as small as possible and only adds IP and User Agent to the cache for the current logged in user.
- No database tables are used. Instead WhosHere uses a proxy model of the User model.
- Proxy models will create migrations but do not affect your database
- Telize lookups are cached for performance
- Using the excellent requests library for better SSL support

##Todo
- Add tests
- Think of other things to add



CHANGELOG.md
#Changelog

##0.1.6.2
- Added 24 hour cache on IP lookups

##0.1.6.1
- Added timeout on Telize lookups to prevent hangups

##0.1.6
- Added tracked request path

##0.1.5
- Added template tags

###0.1.4
- Replaced urllib with requests for better ssl support

###0.1.3
 - Added geolocation with Telize

###0.1.2
 - Added support for better ip lookups with django-ipware

###0.1.1
 - Added Tests
 
##0.1.0
 - Initial commit


django_whoshere/apps.py
from django.apps import AppConfig
from django.conf import settings
from django.core.cache import cache
import requests

# Sets the label name in the admin
try:
    LABEL = settings.WHOSHERE_LABEL
except AttributeError:
    LABEL = 'Active Users'

# Time for a user to be considered idle in seconds
try:
    TIMEOUT = settings.WHOSHERE_TIMEOUT
except AttributeError:
    TIMEOUT = 300

# Cache key prefix
try:
    PREFIX = settings.WHOSHERE_PREFIX
except AttributeError:
    PREFIX = 'whoshere'

# Wether or not to use Telize for ip lookups
try:
    TELIZE = settings.WOSHERE_TELIZE
except AttributeError:
    TELIZE = True

# Check if GeoIp in installed and available
try:
    from django.contrib.gis.geoip import GeoIP

    GEOIP_PATH = settings.GEOIP_PATH
except ImportError:
    GeoIP = None
except AttributeError:
    GeoIP = None

# Check if user-agents is installed
try:
    from user_agents import parse
except ImportError:
    parse = None


class SessionAdminConfig(AppConfig):
    name = 'django_whoshere'
    verbose_name = LABEL


def telize_lookup(ip):
    """
    Queries Telize.com for geolocation.
    Results are held in django cache.
    """
    key = '{}:telize:{}'.format(PREFIX, ip)
    if cache.get(key):
        return cache.get(key)
    try:
        request = requests.get('https://www.telize.com/geoip/{}'.format(ip), timeout=5)
    except requests.exceptions.Timeout:
        location = {'city': 'timeout', 'country': 'timeout'}
    except requests.exceptions.RequestException:
        location = {'city': 'error', 'country': 'error'}
    else:
        data = request.json()
        location = {'city': data.get('city', 'unknown'), 'country': data.get('country', 'unknown')}
        cache.set(key, location, 24 * 60 * 60)
    return location


def get_ip(request):
    """
    Uses ipware to find real ip address if available.
    """
    try:
        import ipware.ip

        return ipware.ip.get_real_ip(request) or request.META.get('REMOTE_ADDR', '')
    except ImportError:
        return request.META.get('REMOTE_ADDR', '')


def get_city(ip):
    """
    Gets city location by ip either with Telize or GeoIP
    """
    if GeoIP:
        key = '{}:geo:{}:city'.format(PREFIX, ip)
        if cache.get(key):
            return cache.get(key)
        geo = GeoIP().city(str(ip))
        if geo and geo['city']:
            cache.set(key, geo['city'], 24 * 60 * 60)
            return geo['city']
        return 'unknown'
    return telize_lookup(ip)['city']


def get_country(ip):
    """
    Gets country location by ip either with Telize or GeoIP
    """
    if GeoIP:
        key = '{}:geo:{}:country'.format(PREFIX, ip)
        if cache.get(key):
            return cache.get(key)
        geo = GeoIP().country(str(ip))
        if geo and geo['country_name']:
            cache.set(key, geo['country_name'], 24 * 60 * 60)
            return geo['country_name']
        return 'unknown'
    return telize_lookup(ip)['country']
